### PR TITLE
DCS-1449 search engine

### PR DESCRIPTION
**Pull request by michalnawrockidj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 1]
```diff
@@ -65,15 +68,54 @@ class PrisonerSearchService(@Autowired private val client: PrisonerSearchApiClie
...
+    listByNameAndDateOfBirth.forEach {
+      map[it.prisonerNumber + "_" + it.lastName + "_" + it.firstName] = PotentialMatch(
+        firstName = it.firstName,
+        lastName = it.lastName,
+        dateOfBirth = it.dateOfBirth,
+        pncNumber = it.pncNumber,
+        prisonNumber = it.prisonerNumber
+      )
     }
+    return map.values.toList()
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:119]: Can't you add all these results to a list? and then use `distinctBy` to remove duplicates based on prison number?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 2]
```diff
@@ -65,15 +68,54 @@ class PrisonerSearchService(@Autowired private val client: PrisonerSearchApiClie
...
+
+    var searchByNameAndDateOfBirth = SearchByNameAndDateOfBirth(
+      firstName = matchPrisonersRequest.firstName,
+      lastName = matchPrisonersRequest.lastName,
+      dateOfBirth = matchPrisonersRequest.dateOfBirth
+    )
+    var listByNameAndDateOfBirth = client.matchPrisonerByNameAndDateOfBirth(searchByNameAndDateOfBirth)
+
+    listByNameAndDateOfBirth.forEach {
+      map[it.prisonerNumber + "_" + it.lastName + "_" + it.firstName] = PotentialMatch(
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:111]: prison number will be unique on its own

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt [hunk 3]
```diff
@@ -65,15 +68,54 @@ class PrisonerSearchService(@Autowired private val client: PrisonerSearchApiClie
     )
 
   fun findPotentialMatch(matchPrisonersRequest: MatchPrisonersRequest): List<PotentialMatch> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:70]: think you're missing test for this?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt (hunk 1 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:119]: What about scenario when I have multiple record back without prison number

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByNameAndDateOfBirth.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.time.LocalDate
+
+data class SearchByNameAndDateOfBirth(
+  val firstName: String,
+  val lastName: String,
+  val dateOfBirth: LocalDate,
+  val pncNumber: String? = null,
+  val croNumber: String? = null,
+  val nomsNumber: String? = null
```
[the remark this answers is not in the document]
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByNameAndDateOfBirth.kt:11]: this is a request send for search only name and DOB are mandatory

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt (hunk 1 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt]: sorted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt (hunk 2 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt]: sorted

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt (hunk 3 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchService.kt:70]: I added test now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/resource/MatchPrisonersResourceTest.kt [hunk 5]
```diff
@@ -35,11 +61,20 @@ class MatchPrisonersResourceTest : IntegrationTestBase() {
       .expectStatus().isOk
       .expectBodyList(PotentialMatch::class.java).returnResult().responseBody
 
-    assertThat(resp.size).isEqualTo(repetitionInfo.currentRepetition - 1)
+    assertThat(resp?.size).isEqualTo(2)
+    assertThat(resp?.get(0)?.lastName).isEqualTo("Larsen")
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/resource/MatchPrisonersResourceTest.kt:65]:
```
with(resp?.get(0)) {
  assertThat(lastName).isEqualTo("Larsen")
}

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**